### PR TITLE
Scrape logs out of containers when integration tests fail

### PR DESCRIPTION
**Pull request by tgeoghegan**
When tests run against interop containers fail, we now copy logs out of them so that they can be rolled up by GitHub Actions and attached to failing test runs as an artifact.

Part of #455

### CODE REVIEW

**Comment by tgeoghegan** [comment]
This seems to be working! I made a change to force test failures. Here's a failing run: https://github.com/divviup/janus/actions/runs/3004476197

We get a link to download the logs artifact. Its contents:
```
./janus_daphne
./janus_daphne/leader-d2d45a824ddd18673c402f61ed2d71fe44c04eed627d9a084720dfd58374d7bc
./janus_daphne/leader-d2d45a824ddd18673c402f61ed2d71fe44c04eed627d9a084720dfd58374d7bc/postgres_stdout.log
./janus_daphne/leader-d2d45a824ddd18673c402f61ed2d71fe44c04eed627d9a084720dfd58374d7bc/postgres_stderr.log
./janus_daphne/leader-d2d45a824ddd18673c402f61ed2d71fe44c04eed627d9a084720dfd58374d7bc/aggregator_stdout.log
./janus_daphne/leader-d2d45a824ddd18673c402f61ed2d71fe44c04eed627d9a084720dfd58374d7bc/aggregator_stderr.log
./janus_daphne/helper-9104509d90cc3fedea7dcbefb98c2ecbf348073163f15d24458673fe581d25f7
./janus_daphne/helper-9104509d90cc3fedea7dcbefb98c2ecbf348073163f15d24458673fe581d25f7/stdout.log
./janus_daphne/helper-9104509d90cc3fedea7dcbefb98c2ecbf348073163f15d24458673fe581d25f7/stderr.log
./daphne_janus
./daphne_janus/leader-ce359b3f0b57462955a524afe420d06ccaef55fe98f6cbf840b6a42b42dbf285
./daphne_janus/leader-ce359b3f0b57462955a524afe420d06ccaef55fe98f6cbf840b6a42b42dbf285/stdout.log
./daphne_janus/leader-ce359b3f0b57462955a524afe420d06ccaef55fe98f6cbf840b6a42b42dbf285/stderr.log
./daphne_janus/helper-e3dd356a2babf57492b8139670885a4201bca6ce030f2b8625a7dd432d82bef7
./daphne_janus/helper-e3dd356a2babf57492b8139670885a4201bca6ce030f2b8625a7dd432d82bef7/postgres_stdout.log
./daphne_janus/helper-e3dd356a2babf57492b8139670885a4201bca6ce030f2b8625a7dd432d82bef7/postgres_stderr.log
./daphne_janus/helper-e3dd356a2babf57492b8139670885a4201bca6ce030f2b8625a7dd432d82bef7/aggregator_stdout.log
./daphne_janus/helper-e3dd356a2babf57492b8139670885a4201bca6ce030f2b8625a7dd432d82bef7/aggregator_stderr.log
```
The top level path is the name of the failing test (or rather, a label chosen by the test and passed into the log fetcher function). Then we create a directory named after the role being played and the Docker container ID, inside of which are the stdout and stderrs of the container's processes.

File: integration_tests/tests/common/mod.rs [hunk 1]
```diff
@@ -88,11 +105,33 @@ pub fn translate_url_for_external_access(url: &Url, external_port: u16) -> Url {
...
+    Leader: CopyLogs,
+    Fut: Future<Output = Result<()>>,
+    Test: FnMut() -> Fut,
+>(
+    test_name: &str,
+    helper: &Helper,
+    leader: &Leader,
+    mut test: Test,
+) {
+    let result = test().await;
```
Comment (divergentdave) [inline integration_tests/tests/common/mod.rs:120]: We should probably use `std::panic::catch_unwind()` here, both to ensure we emit logs on any panics we miss, and to make writing tests in this context more ergonomic. (We can use `std::panic::panic_any()` to propagate resulting outer errors)

**Review by divergentdave — APPROVED** [review]

File: integration_tests/tests/common/mod.rs (hunk 1 above)
Comment (tgeoghegan) [reply integration_tests/tests/common/mod.rs:120]: I tried doing that, but `std::panic::catch_unwind` can't be passed an async closure. The async equivalent is [`futures::future::FutureExt::catch_unwind`](https://docs.rs/futures/latest/futures/future/trait.FutureExt.html#method.catch_unwind), but it requires `UnwindSafe`, which the closure we want to test is decidedly not, and so I backed off from this approach and instead made `submit_measurements_and_verify_aggregate` fallible.

Comment (divergentdave) [reply integration_tests/tests/common/mod.rs:120]: Ah right, of course. There are future wrappers that apply this to poll, FWIW: https://docs.rs/futures/latest/futures/future/struct.CatchUnwind.html

**Pull request merged**